cli/src/commands/fix.rs

```diff
@@ -48,106 +69,260 @@ pub(crate) fn cmd_fix(
     }
 
-    let formatted = format_files(tx.repo().store().as_ref(), &file_ids)?;
+    // Run the configured tool on all of the chosen inputs.
+    // TODO: Support configuration of multiple tools and which files they affect.
+    let tool_command: CommandNameAndArgs = command
+        .settings()
+        .config()
+        .get("fix.tool-command")
+        .map_err(|err| config_error_with_message("Invalid `fix.tool-command`", err))?;
+    let fixed_file_ids = fix_file_ids(
+        tx.repo().store().as_ref(),
+        &tool_command,
+        &unique_tool_inputs,
+    )?;
 
+    // Substitute the fixed file IDs into all of the affected commits. Currently,
+    // fixes cannot delete or rename files, change the executable bit, or modify
+    // other parts of the commit like the description.
+    let mut num_checked_commits = 0;
+    let mut num_fixed_commits = 0;
     tx.mut_repo().transform_descendants(
         command.settings(),
         root_commits.iter().ids().cloned().collect_vec(),
         |mut rewriter| {
-            let paths = commit_paths.get(rewriter.old_commit().id()).unwrap();
+            // TODO: Build the trees in parallel before `transform_descendants()` and only
+            // keep the tree IDs in memory, so we can pass them to the rewriter.
+            let tool_inputs_map = commit_tool_inputs_maps
+                .get(rewriter.old_commit().id())
+                .unwrap();
             let old_tree = rewriter.old_commit().tree()?;
             let mut tree_builder = MergedTreeBuilder::new(old_tree.id().clone());
-            for path in paths {
-                let old_value = old_tree.path_value(path);
+            let mut changes = 0;
+            for (repo_path, tool_inputs) in tool_inputs_map {
+                let old_value = old_tree.path_value(repo_path)?;
                 let new_value = old_value.map(|old_term| {
                     if let Some(TreeValue::File { id, executable }) = old_term {
-                        if let Some(new_id) = formatted.get(&(path, id)) {
-                            Some(TreeValue::File {
-                                id: new_id.clone(),
-                                executable: *executable,
-                            })
-                        } else {
-                            old_term.clone()
+                        // We could use a map from N ToolInputs to file IDs, but N is likely small.
+                        for tool_input in tool_inputs {
+                            if tool_input.file_id == *id {
+                                if let Some(new_id) = fixed_file_ids.get(&tool_input) {
+                                    return Some(TreeValue::File {
+                                        id: new_id.clone(),
+                                        executable: *executable,
+                                    });
+                                }
+                            }
                         }
```

Comment:
It's simpler to just create a new `ToolInput` from `repo_path` and `id`. That avoids the loop. We can also simplify `tool_inputs_map` back to the `commit_paths` it was before (but I agree with changing the `Vec<RepoPath>` to a `HashSet<RepoPath>` to avoid duplicates)

Reply:
Oh, I attempted that last night before I wrote this comment, so I should just share that with you. It's in e93c2d6c35f0e2f2b5b5c14995daf4762e5e73a0.

Reply:
I've updated that in branch push-yowrlmvyxssp

Reply:
Squashed your change, with a couple of comment tweaks and renamed `tool_inputs_map` to `repo_paths`.